File: server/src/Marshaller/ClientPayload.java
package src.marshaller;

import java.util.Date;

/**
 * Class to store the payload for the client's request specific for a service
 */
public class ClientPayload {
    /**
     * The offset to start reading from (only for Service 1 and 2)
     */
    private int offset;
    /**
     * The number of bytes to read (only for Service 1)
     */
    private int numBytes;
    /**
     * The bytes to insert (only for Service 2)
     */
    private byte[] bytesToInsert;
    /**
     * The expiry date for monitoring (only for Service 3)
     */
    private Date expiryDate;

    /**
     * Constructor for the payload for Service 1
     * @param offset    offset to start reading from
     * @param numBytes  number of bytes to read
     */
    public ClientPayload(int offset, int numBytes) {
        this.offset = offset;
        this.numBytes = numBytes;
    }

    /**
     * Constructor for the payload for Service 2
     * @param offset            offset to start reading from
     * @param bytesToInsert     bytes to insert
     */
    public ClientPayload(int offset, byte[] bytesToInsert) {
        this.offset = offset;
        this.bytesToInsert = bytesToInsert;
    }

    /**
     * Constructor for the payload for Service 3
     * @param expiryDate    expiry date for monitoring
     */
    public ClientPayload(Date expiryDate) {
        this.expiryDate = expiryDate;
    }

    /**
     * Default constructor for service 4 and 5
     */
    public ClientPayload() {}

    /**
     * Getter method for the offset
     * @return  The offset
     */
    public int getOffset() {
        return offset;
    }

    /**
     * Getter method for the number of bytes
     * @return  The number of bytes
     */
    public int getNumBytes() {
        return numBytes;
    }

    /**
     * Getter method for the bytes to insert
     * @return  The bytes to insert
     */
    public byte[] getBytesToInsert() {
        return bytesToInsert;
    }

    /**
     * Getter method for the expiry date
     * @return  The expiry date
     */
    public Date getExpiryDate() {
        return expiryDate;
    }
}

File: client/src/ClientPayload.cpp
/**
 * @file ClientPayload.cpp
 * @brief Implementation file for the ClientPayload class.
 */

#include "ClientPayload.h"

// Constructor
ClientPayload::ClientPayload() {}

/**
 * @brief Constructor for ClientPayload class.
 * @param expirationTime The expiration time value to set.
 */
ClientPayload::ClientPayload(unsigned long long expirationTime)
    : expirationTime(expirationTime) {}

/**
 * @brief Constructor for ClientPayload class.
 * @param offset The offset value to set.
 * @param numBytes The number of bytes to set.
 */
ClientPayload::ClientPayload(int offset, int numBytes)
    : offset(offset), numBytes(numBytes) {}

/**
 * @brief Constructor for ClientPayload class.
 * @param offset The offset value to set.
 * @param input The input string to set.
 */
ClientPayload::ClientPayload(int offset, const std::string& input)
    : offset(offset), input(input) {}

/**
 * @brief Get the offset value of the client payload.
 * @return The offset value.
 */
int ClientPayload::getOffset() const { return offset; }

/**
 * @brief Get the number of bytes of the client payload.
 * @return The number of bytes.
 */
int ClientPayload::getNumBytes() const { return numBytes; }

/**
 * @brief Get the expiration time of the client payload.
 * @return The expiration time.
 */
unsigned long long ClientPayload::getExpirationTime() const {
    return expirationTime;
}

/**
 * @brief Get the input string of the client payload.
 * @return The input string.
 */
const std::string& ClientPayload::getInput() const { return input; }

// Setters

/**
 * @brief Set the offset value of the client payload.
 * @param offset The offset value to set.
 */
void ClientPayload::setOffset(int offset) { this->offset = offset; }

/**
 * @brief Set the number of bytes of the client payload.
 * @param numBytes The number of bytes to set.
 */
void ClientPayload::setNumBytes(int numBytes) { this->numBytes = numBytes; }

/**
 * @brief Set the expiration time of the client payload.
 * @param expirationTime The expiration time to set.
 */
void ClientPayload::setExpirationTime(unsigned long long expirationTime) {
    this->expirationTime = expirationTime;
}

/**
 * @brief Set the input string of the client payload.
 * @param input The input string to set.
 */
void ClientPayload::setInput(const std::string& input) { this->input = input; }